mongoengine/base/fields.py

```diff
@@ -103,9 +104,12 @@ def __get__(self, instance, owner):
                     value = self.default() if callable(self.default) else self.default
                 else:
                     value = self.to_python(db_value)
+                    if isinstance(value, DocumentProxy):
+                        value._set_parent_ref(instance, name)
```

Comment:
I'd suggest to avoid holding onto the parent instance as this might cause memory leaks. Store just the class (name or reference), and the instance pk.